File: TestLeetCode/Tree/LeastCommonAncestor.cs
public class LowestCommonAncestorSolution {
    public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q)
    {
        if (root == null) { return null; }
        if (root == p || root == q) { return root; }
        
        var left = LowestCommonAncestor(root.left, p, q);
        var right = LowestCommonAncestor(root.right, p, q);
        if (left != null && right != null)
        {
            return root;
        }
        
        return left == null ? right : left;
    }
}

File: python/tree/least_common_ancestor.py
from python.common import TreeNode


class LowestCommonAncestor:
    def lowest_common_ancestor(self, root: TreeNode | None, left: TreeNode, right: TreeNode) -> TreeNode | None:
        if not root or root is left or root is right: return root
        found_left = self.lowest_common_ancestor(root.left, left, right); found_right = self.lowest_common_ancestor(root.right, left, right)
        return root if found_left and found_right else found_left or found_right
